modules/kvformat/kv-parser.c

```diff
@@ -75,16 +76,22 @@ kv_parser_set_stray_words_value_name(LogParser *s, const gchar *value_name)
 }
 
 static const gchar *
-_get_formatted_key(KVParser *self, const gchar *key)
+_get_formatted_key_with_prefix(KVParser *self, const gchar *key, GString *formatted_key)
+{
+  if (formatted_key->len > 0)
+    g_string_truncate(formatted_key, self->prefix_len);
+  else
+    g_string_assign(formatted_key, self->prefix);
```

Comment:
I suggest doing this assignment once in the beginning of _process before the while loop. This function can unconditionally truncate then append without the branching.

Reply:
Oh I did not realise for some reason we dropped precalculating the formatted key getter function. I liked the idea, we are doing something very similar ing csvparser.c

```
  GString *key_scratch = scratch_buffers_alloc();
  if (self->prefix)
    g_string_assign(key_scratch, self->prefix);

  key_formatter_t _key_formatter = dispatch_key_formatter(self->prefix);
  while (csv_scanner_scan_next(&scanner))
    {

      log_msg_set_value_by_name(msg,
                                _key_formatter(key_scratch, csv_scanner_get_current_name(&scanner), self->prefix_len),
                                csv_scanner_get_current_value(&scanner),
                                csv_scanner_get_current_value_len(&scanner));
    }
```

@szemere you can ignore my previous comment.

Reply:
@furiel, @szemere: my only problem was that we are storing the formatter in KVParser and the logic around setting up the formatter and prefix/formattes ket was splitted. So the formatter itself can be used as in case of csvparser. I also want to highlight that in a language that supports lambdas it can be expressed in a more elegant way(in our case the signature is ‘polluted’ due to the limitations of C programming language, but this is just a matter of taste whether we use fptrs or another solution).